proptest/src/num/float_samplers.rs

```diff
@@ -0,0 +1,318 @@
+                a - last_float(a)
+            }
+
+            #[derive(Copy, Clone, Debug)]
+            pub(crate) struct $wrapper($typ);
+
+            impl From<$typ> for $wrapper {
+                fn from(x: $typ) -> Self {
+                    $wrapper(x)
+                }
+            }
+            impl From<$wrapper> for $typ {
+                fn from(x: $wrapper) -> Self {
+                    x.0
+                }
+            }
+
+            #[derive(Clone, Copy, Debug)]
+            pub(crate) struct FloatUniform {
+                uniform: Uniform<$int_typ>,
+                values: SampleValueCollection,
+            }
+
+            impl UniformSampler for FloatUniform {
+
+                type X = $wrapper;
+
+                fn new<B1, B2>(low: B1, high: B2) -> Self
+                where
+                    B1: SampleBorrow<Self::X> + Sized,
+                    B2: SampleBorrow<Self::X> + Sized,
+                {
+                    let low = low.borrow().0;
+                    let high = high.borrow().0;
+
+                    let values = SampleValueCollection::new(low, last_float(high));
+
+                    FloatUniform {
+                        uniform: Uniform::new(0, values.count),
+                        values,
+                    }
+                }
+
+                fn new_inclusive<B1, B2>(low: B1, high: B2) -> Self
+                where
+                    B1: SampleBorrow<Self::X> + Sized,
+                    B2: SampleBorrow<Self::X> + Sized,
+                {
+                    let low = low.borrow().0;
+                    let high = high.borrow().0;
+
+                    let values = SampleValueCollection::new_inclusive(low, high);
+
+                    FloatUniform {
+                        uniform: Uniform::new(0, values.count),
+                        values,
+                    }
+                }
+
+                fn sample<R: Rng + ?Sized>(&self, rng: &mut R) -> Self::X {
+                    $wrapper(self.values.get(self.uniform.sample(rng)))
+                }
+            }
+
+            impl SampleUniform for $wrapper {
+                type Sampler = FloatUniform;
+            }
+
+            #[derive(Clone, Copy, Debug)]
+            struct SampleValueCollection {
+                start: $typ,
+                end: $typ,
+                step: $typ,
+                count: $int_typ,
+            }
+
+            // Values greater than MAX_PRECISE_INT may be rounded when converted to float.
+            const MAX_PRECISE_INT: $int_typ =
+                (2 as $int_typ).pow($typ::MANTISSA_DIGITS);
+
+            // The collection of sample values that may be generated by UniformF32U.
+            impl SampleValueCollection {
+                fn new(low: $typ, high: $typ) -> Self {
+                    let high_incl = last_float(high);
+
+                    Self::new_inclusive(low, high_incl)
+                }
```

Comment:
with above comment: propose removing this function and only have `new_inclusive` to prevent possibilities of calculating `last_float(high)` twice.